### ISSUE TITLE
 How to run programs with different configurations on different GPUs?

### ISSUE THREAD

**Issue by zhiwenwang20122302** [ultralytics/ultralytics#18352]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train, Multi-GPU

### Bug

Dear Author,  

Hello! When I train YOLOv8 on a single GPU, one epoch takes about one minute. However, when I run four different configuration programs on four GPUs simultaneously, the runtime for each program increases by 3-4 times. It is clear that these programs are interfering with each other.  

What should I do to ensure these programs do not interfere with each other? Thank you in advance!  

Best regards.

### Environment

Ultralytics 8.3.49 🚀 Python-3.9.21 torch-1.13.1+cu116 CUDA:0 (NVIDIA GeForce RTX 3090, 24260MiB)
Setup complete ✅ (64 CPUs, 62.5 GB RAM, 53.9/97.9 GB disk)

OS                  Linux-5.4.0-200-generic-x86_64-with-glibc2.31
Environment         Linux
Python              3.9.21
Install             git
RAM                 62.50 GB
Disk                53.9/97.9 GB
CPU                 Intel Xeon Silver 4314 2.40GHz
CPU count           64
GPU                 NVIDIA GeForce RTX 3090, 24260MiB
GPU count           4
CUDA                11.6

numpy               ✅ 1.26.0>=1.23.0
numpy               ✅ 1.26.0<2.0.0; sys_platform == "darwin"
matplotlib          ✅ 3.9.4>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 11.0.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.13.1>=1.4.1
torch               ✅ 1.13.1+cu116>=1.8.0
torch               ✅ 1.13.1+cu116!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.14.1+cu116>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 6.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.13>=2.0.0
{'OS': 'Linux-5.4.0-200-generic-x86_64-with-glibc2.31', 'Environment': 'Linux', 'Python': '3.9.21', 'Install': 'git', 'RAM': '62.50 GB', 'Disk': '53.9/97.9 GB', 'CPU': 'Intel Xeon Silver 4314 2.40GHz', 'CPU count': 64, 'GPU': 'NVIDIA GeForce RTX 3090, 24260MiB', 'GPU count': 4, 'CUDA': '11.6', 'Package Info': {'numpy': '✅ 1.26.0<2.0.0; sys_platform == "darwin"', 'matplotlib': '✅ 3.9.4>=3.3.0', 'opencv-python': '✅ 4.10.0.84>=4.6.0', 'pillow': '✅ 11.0.0>=7.1.2', 'pyyaml': '✅ 6.0.2>=5.3.1', 'requests': '✅ 2.32.3>=2.23.0', 'scipy': '✅ 1.13.1>=1.4.1', 'torch': '✅ 1.13.1+cu116!=2.4.0,>=1.8.0; sys_platform == "win32"', 'torchvision': '✅ 0.14.1+cu116>=0.9.0', 'tqdm': '✅ 4.67.1>=4.64.0', 'psutil': '✅ 6.0.0', 'py-cpuinfo': '✅ 9.0.0', 'pandas': '✅ 2.2.3>=1.1.4', 'seaborn': '✅ 0.13.2>=0.11.0', 'ultralytics-thop': '✅ 2.0.13>=2.0.0'}}

### Minimal Reproducible Example

import argparse
from ultralytics import YOLO
import os

parser = argparse.ArgumentParser(description='.')
parser.add_argument('--config', type=str, default='default')
parser.add_argument('--bs', type=int, default=8)
parser.add_argument('--device', type=str, default='0')
args = parser.parse_args()

if __name__ == "__main__":
    epochs = 100
    imgsz = 1280
    workers =  8
    
    device = list(map(int, args.device.split(',')))  

    batch=args.bs
    model = YOLO(args.config, task='detect')

    nw = min(workers, os.cpu_count(), batch)
    results = model.train(data='data.yaml', 
                          epochs=epochs, 
                          imgsz=imgsz, 
                          amp=False,
                          batch=batch, 
                          workers=nw, 
                          device=device)

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @zhiwenwang20122302, thank you for your interest in Ultralytics 🚀! We appreciate you taking the time to share your detailed report and environment setup.

We recommend reviewing the [Docs](https://docs.ultralytics.com) for valuable information, including [Training Tips](https://docs.ultralytics.com/guides/model-training-tips/) and [Multi-GPU Training](https://docs.ultralytics.com/usage/cfg/#multi-gpu-training), which may help resolve your issue. Many common questions or scenarios may already have solutions outlined there.

If this is a 🐛 Bug Report, we kindly ask you to provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) (MRE) that isolates the issue. Your code snippet is already quite helpful—thank you! However, to ensure we can effectively debug the interference between your processes, please confirm:
1. The configurations being used for each of the four programs.
2. Clear instructions to reproduce the issue, including any specific customizations made to your `data.yaml` or training pipeline.
3. Logs from your training processes, if available.

Additionally, you can try upgrading to the latest `ultralytics` package, as your current version, `8.3.49`, may not contain the latest updates and fixes. To upgrade:
```bash
pip install -U ultralytics
```
Ensure you're running this in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/).

### Verified Environments

For best performance and reproducibility, confirm you are running YOLO in one of our verified environments:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: See our [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: See our [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: See our [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

If you suspect this is a resource contention issue between GPUs, you could experiment with the following:
- Utilize the `CUDA_VISIBLE_DEVICES` environment variable to explicitly allocate GPUs to specific processes.
- Ensure there is no CPU bottleneck by monitoring CPU utilization with tools like `htop` or `nvidia-smi dmon` while your programs are running.
- Experiment with reducing batch size or workers per process to reduce potential contention.

### Community Support

For additional assistance, join our vibrant community:
- Real-time chat: [Discord](https://discord.com/invite/ultralytics) 🎧
- Discussions: [Discourse](https://community.ultralytics.com)
- Knowledge sharing: [Subreddit](https://reddit.com/r/Ultralytics) 

### CI Status

Our [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) verifies all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) daily and on every commit. <a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="CI Status"></a>

This is an automated response to assist you efficiently. An Ultralytics engineer will review and provide additional insights soon. Thank you for your patience! 😊

**Comment by zhiwenwang20122302** [comment]
In the official documentation, I can see how multi-GPU parallel processing is handled, but that's not what I want. My question is not about how to implement distributed parallelism for a single program. Instead, it's about how to ensure that when running independent processes with different configuration files on separate GPUs, the training speed does not interfere with each other.

**Comment by Y-T-G** [comment]
Your environment log only shows 1 GPU.

**Comment by Y-T-G** [comment]
You can reduce the workers for each. Otherwise the CPU resources are being shared.

**Comment by Y-T-G** [comment]
Can you post `nvidia-smi` logs when 4 trainings are running?
